src/app/core/services/models/part-entity.service.ts

```diff
@@ -102,4 +102,15 @@ export class PartEntityService {
       pieces: [...part.pieces, piece],
     }
   }
+
+  public replacePiece(part: Part, replacedPieceId: string, newPiece: Piece): Part {
+    const replacedPieceIndex: number = part.pieces.findIndex(piece => piece.id === replacedPieceId)
+    if (replacedPieceIndex >= 0) {
+      part.pieces.splice(replacedPieceIndex, 1)
+    }
+    return {
+      ...part,
+      pieces: [...part.pieces, newPiece],
+    }
```

Comment:
`Array.splice` mutates the referenced object, which we should avoid. An alternative to this is to use `Array.toSpliced` or use `Array.filter`. E.g.:

```typescript
return {
  ..part,
  pieces: [
    ...part.pieces.filter(piece => piece.id !== replacedPiece.id),
    newPiece
  ]
}
```
or
```typescript
const replacedPieceIndex: number = part.pieces.findIndex(piece => piece.id === replacedPiece.id)
if (replacedPieceIndex < 0) {
  return part
}
return {
  ...part,
  pieces: [...part.pieces.toSpliced(replacedPieceIndex, 1), newPiece],
}
```

Reply:
A very good point and an elegant solution. I made use of filtering! Thanks!